### ISSUE TITLE
[Feature] Optionally enable alarm-style reminders

### ISSUE THREAD

**Issue by Shadow53** [TechbeeAT/jtxBoard#809]
**Is your feature request related to a problem? Please describe.**

When I set an alarm for a todo item, I want it to be a reminder to start doing the thing. Yet I tend to miss notifications on my phone if I am not staring at my screen when they appear. 

**Describe the solution you'd like**

An option -- either in settings or per-todo-alarm -- to enable "full screen" noisy reminders, similar to when an alarm or timer goes off in the default clock app.

- The window/activity covers the entire screen
- An optional alarm tone and/or vibration starts playing
- It does not go away until dismissed

**Describe alternatives you've considered**

Setting a manual alarm would theoretically work, but I want the reminder to be synchronized across multiple devices, which would not happen manually.

**Additional context**

Possible actions I see being useful from the alarm screen -- these are nice to have, but not as important as the full-screen behavior (except a way to dismiss):

- Dismiss/Acknowledge (required)
- Snooze
- Start task (updates task status)
- Complete task (updates task status)
- Open task (opens app to task details activity)
    - From here, could manually mark started/completed

**Comment by patrickunterwegs** [comment]
Hi @Shadow53 thank you for your feature request.
I am not fully convinced by the idea. I mean, I get what you mean, but as you say, such a full screen notification is something you'd expect from an alarm clock, but from a tasks app? But nevertheless, I'll check what can be done and I'll comment again when I get deeper into this topic.

**Comment by Shadow53** [comment]
I appreciate you looking into this!

I am not an Android app developer, so I'm not familiar with the design patterns, but I went looking through the Android developer docs and I think I found what I am asking for.

What I would like is to be able to set the jtxBoard reminder notification channel to ["Urgent" importance](https://developer.android.com/develop/ui/views/notifications/channels#importance) as, at least for me, the reminders are [time sensitive](https://developer.android.com/develop/ui/views/notifications/time-sensitive). According to the previous link, part of the feature request entails setting in the app to make the notification also send a full-screen intent (the "alarm" screen).

If the full-screen part still doesn't make sense from a design perspective, I would be content with increasing the priority of the notifications so they show up at the top of the notification drawer instead of in the middle.

**Comment by patrickunterwegs** [comment]
Yes, for the alarm filling the screen I'd say the fullscreen intent is anyway to go. 
I was a bit hesitant adding another setting, but I guess it would still be best. 

The importance of the notification channel is already set to high:

https://github.com/TechbeeAT/jtxBoard/blob/89505cdd1aa537a37df2f378e15b491eef243224/app/src/main/java/at/techbee/jtx/MainActivity2.kt#L291C13-L291C25

Do you see jtx Board notifications being shown differently? Could you please also check the system settings that you have for jtx Board?

For the alarm screen itself I just need to find the time and energy. I just started a new job and it's just requires a lot of attention for the start 😉

**Comment by patrickunterwegs** [comment]
Ok, I See that also IMPORTANCE_MAX is possible for the notification channel, I'm settings this one now.

**Comment by Shadow53** [comment]
I see that you've addressed this already, but for future reference of anyone coming across this issue:

> The importance of the notification channel is already set to high... Do you see jtx Board notifications being shown differently? Could you please also check the system settings that you have for jtx Board?

My notifications are split into three sections. The topmost has notifications from messaging apps like Discord and Element. The second has most other notifications like email, news apps, etc. The third is labeled "Silent" and appears to be mostly persistent notifications.

The "Notifications" settings for the app allow me to choose whether notifications make sound or ignore Do Not Disturb, but not priority. That said, it is set to as noisy/intrusive as it can be.

**Comment by patrickunterwegs** [comment]
Hey @Shadow53 , 
so finally I found the time to implement this feature. It was a bit more complex than expected. 

Basically the app can also request the permission for USE_FULL_SCREEN_INTENT now, this would allow the notification to be a full-screen activity. I added an option in the settings (which is also grouped now in specific alarms settings), where you can activate full-screen alarms. The app would redirect you to the settings, if the permission for the full-screen notifications is deactivated in the system settings. 

I am preparing a new version now for the Google Play beta-channel. If you have some time, it would be nice if you could join the beta-channel there and give some early feedback! It should be available by tomorrow. The version with the enhancement will be 2.6.1-beta01!

**Comment by Shadow53** [comment]
Hey @patrickunterwegs, thanks for taking this on! I'm mostly happy with it, just a few tweaks, I think:

First, the settings option says "Shows alarms in full screen when the device is locked". From that, I expected something like the alarm app where it turns the screen on and shows the full screen activity *over* the lock screen. Instead, the screen turned on, showing the lock screen, but with no obvious indication of what caused the screen to turn on. When I unlocked my phone, *then* the full screen view appeared. I'm wondering if that was intentional, if it's a limitation in Android, or something else.

Second, the full screen view does not adapt to the user's light/dark preferences. At least, I got a view with dark text on a dark background:

![Screenshot_20231009-214819](https://github.com/TechbeeAT/jtxBoard/assets/10555070/cacd46b0-faa0-43ec-9512-4a68b19f0b27)

That said, I do like the heads-up notification that appears when I'm using my phone when the alarm happens, and how it doesn't automatically go away. I couldn't figure out how to *manually* dismiss it without choosing one of the button options. Which brings me to the last bit of constructive feedback, which is regarding those options:

- I think "Done" would be more clear as "(Mark) Complete", as I initially hit "Done" thinking it meant "Acknowledge this notification and make it go away"
- A "+5 minute" option would be useful for those times where I need to finish up what I am doing but the reminder soon. 1 hour and 1 day seem too long for that use case. I think my preference would be 5 minutes and 1 hour, and make 1 day require opening the app and manually editing the alarm.
- I think that the notification contents are the same as before this was implemented, and I am sure other people like the delay options currently available, so take the above bullets as data points and not me saying the app must change -- though I do feel somewhat strongly about changing "Done".

**Comment by patrickunterwegs** [comment]
Thank you @Shadow53 for your feedback!

So let's take it one by one: 

> First, the settings option says "Shows alarms in full screen when the device is locked". From that, I expected something like the alarm app where it turns the screen on and shows the full screen activity over the lock screen. Instead, the screen turned on, showing the lock screen, but with no obvious indication of what caused the screen to turn on. When I unlocked my phone, then the full screen view appeared. I'm wondering if that was intentional, if it's a limitation in Android, or something else.

I will have another look, but basically it was intentional because of the restrictions in Android. The system is very restrictive when it comes to open a fullscreen activity when the actual app is not open. You can only tell the system, that the alarm should be a fullscreen activity, but then it would still be up to the system to decide what to do with it - ie. to either show only the notification or show the fullscreen activity... I tried some methods to ensure the fullscreen activity, but they were all blocked by the system.

> Second, the full screen view does not adapt to the user's light/dark preferences. At least, I got a view with dark text on a dark background

Thanks, that's a bug (actually I missed putting the parameters), it's fixed now in #944 , I'll upload a new version soon

> I couldn't figure out how to manually dismiss it without choosing one of the button options. 

Actually I put a little X on the top right corner to close the full screen alarm ;-) Do you think this should be more obvious? First I had it as a button below, but then it looked kind of messy with too many options...

> I think "Done" would be more clear as "(Mark) Complete", as I initially hit "Done" thinking it meant "Acknowledge this notification and make it go away"

Hm, I get what you mean. I chose "Done" for the action in the notification because it's short and fits easily in the notification bar. I can imagine putting "Mark as completed" or "Mark as done" or just "Completed" instead of "Done", but for consistency it should be called the same in the notification and the full screen activity... I'm not sure what would be best here...

> A "+5 minute" option would be useful for those times where I need to finish up what I am doing but the reminder soon. 1 hour and 1 day seem too long for that use case. I think my preference would be 5 minutes and 1 hour, and make 1 day require opening the app and manually editing the alarm.

Hm, yes, I  get it... I'm thinking of making a very flexible option where you can choose minutes or days for a delay. But also here I have second thoughts as it might be better to keep it consistent with the notification...
